### PR TITLE
Some mesontest related improvements.

### PR DESCRIPTION
**Pull request by trhd**
Main change being that the ninja target "test" and mesontest now behave more similarly.

### CODE REVIEW

File: mesontest.py [hunk 1]
```diff
@@ -101,9 +101,10 @@ def get_log(self):
         if self.cmd is None:
             res += 'NONE\n'
         else:
-            res += "\n%s %s\n" %(' '.join(
-                ["%s='%s'" % (k, v) for k, v in self.env.items()]),
-                ' ' .join(self.cmd))
+            res += "%s%s\n" % (
+                    ''.join(["%s='%s' " % (k, v) for k, v in self.env.items()]),
```
Comment (ignatenkobrain) [inline mesontest.py]: __[flake8]__
* [E126] continuation line over-indented for hanging indent

File: mesontest.py [hunk 2]
```diff
@@ -101,9 +101,10 @@ def get_log(self):
...
+                    ' ' .join(self.cmd)
+                )
```
Comment (ignatenkobrain) [inline mesontest.py]: __[flake8]__
* [E121] continuation line under-indented for hanging indent

File: mesontest.py [hunk 3]
```diff
@@ -284,189 +284,211 @@ def print_stats(self, numlen, tests, name, result, i, logfile, jsonlogfile):
...
-        self.is_run = True
-        logfilename = self.run_tests(self.options.logbase)
+        if logfile:
+            logfile.write(result_str)
+        if jsonlogfile:
+            write_json_log(jsonlogfile, name, result)
+
+    def print_summary(self, logfile, jsonlogfile):
+        msg = 'Test summary: %d OK, %d FAIL, %d SKIP, %d TIMEOUT' \
+                % (self.success_count, self.fail_count, self.skip_count, self.timeout_count)
```
Comment (ignatenkobrain) [inline mesontest.py]: __[flake8]__
* [E126] continuation line over-indented for hanging indent

File: mesontest.py [hunk 4]
```diff
@@ -284,189 +284,211 @@ def print_stats(self, numlen, tests, name, result, i, logfile, jsonlogfile):
...
-            return
-        numlen = len('%d' % len(tests))
-        executor = conc.ThreadPoolExecutor(max_workers=self.options.num_processes)
-        futures = []
-        filtered_tests = filter_tests(self.options.suite, tests)
-
-        jsonlogfile = None
-        logfile = None
+
+        jsonlogfile =  open(jsonlogfilename, 'w')
```
Comment (ignatenkobrain) [inline mesontest.py]: __[flake8]__
* [E222] multiple spaces after operator

File: mesontest.py [hunk 5]
```diff
@@ -284,189 +284,211 @@ def print_stats(self, numlen, tests, name, result, i, logfile, jsonlogfile):
...
+        logfile = open(logfilename, 'w')
+
+        logfile.write('Log of Meson test suite run on %s.\n\n' %
+                    datetime.datetime.now().isoformat())
```
Comment (ignatenkobrain) [inline mesontest.py]: __[flake8]__
* [E128] continuation line under-indented for visual indent

File: mesontest.py [hunk 6]
```diff
@@ -284,189 +284,211 @@ def print_stats(self, numlen, tests, name, result, i, logfile, jsonlogfile):
...
+
+        return (logfile, logfilename, jsonlogfile, jsonlogfilename)
+
+    def get_wrapper(self):
+        if self.options.gdb:
+            wrap = ['gdb', '--quiet', '--nh' ]
```
Comment (ignatenkobrain) [inline mesontest.py]: __[flake8]__
* [E202] whitespace before ']'

File: mesontest.py [hunk 7]
```diff
@@ -284,189 +284,211 @@ def print_stats(self, numlen, tests, name, result, i, logfile, jsonlogfile):
...
+
+                    if not test.is_parallel or self.options.gdb:
+                        self.drain_futures(futures, logfile, jsonlogfile)
                         futures = []
                         res = self.run_single_test(wrap, test)
-                        if not self.options.verbose:
-                            self.print_stats(numlen, filtered_tests, visible_name, res, i,
-                                            logfile, jsonlogfile)
+                        self.print_stats(numlen, tests, visible_name, res, i,
+                                logfile, jsonlogfile)
```
Comment (ignatenkobrain) [inline mesontest.py]: __[flake8]__
* [E128] continuation line under-indented for visual indent

File: mesontest.py [hunk 8]
```diff
@@ -284,189 +284,211 @@ def print_stats(self, numlen, tests, name, result, i, logfile, jsonlogfile):
...
                     else:
+                        if not executor:
+                            executor = conc.ThreadPoolExecutor(max_workers=self.options.num_processes)
                         f = executor.submit(self.run_single_test, wrap, test)
-                        if not self.options.verbose:
-                            futures.append((f, numlen, filtered_tests, visible_name, i,
-                                            logfile, jsonlogfile))
+                        futures.append((f, numlen, tests, visible_name, i,
+                            logfile, jsonlogfile))
```
Comment (ignatenkobrain) [inline mesontest.py]: __[flake8]__
* [E128] continuation line under-indented for visual indent

File: mesontest.py [hunk 9]
```diff
@@ -284,189 +284,211 @@ def print_stats(self, numlen, tests, name, result, i, logfile, jsonlogfile):
...
-
-def filter_tests(suite, tests):
-    if suite is None:
-        return tests
-    return [x for x in tests if suite in x.suite]
-
+        self.run_tests(tests)
+        return self.fail_count
+
+def list_tests(th):
```
Comment (ignatenkobrain) [inline mesontest.py]: __[flake8]__
* [E302] expected 2 blank lines, found 1

File: mesonbuild/interpreter.py [hunk 10]
```diff
@@ -1998,6 +1998,8 @@ def add_test(self, node, args, kwargs, is_base_test):
                     s = '.' + s
                 newsuite.append(self.subproject.replace(' ', '_').replace('.', '_') + s)
             suite = newsuite
+        else:
+            suite = [ self.build.project_name ]
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E201] whitespace after '['
* [E202] whitespace before ']'

**Comment by jpakkane** [comment]
Giving tests a suite based on the name of the project seems a bit iffy, because people may use that as a suite name already. For project `foo` you would have tests of the core that would be tagged `foo` and tests for non-core (such as helper binaries) would be empty.

The functionality itself is useful and something we want to support. Maybe we can come up with a different UI for it so there is less of a chance of clashes.

**Comment by trhd** [comment]
The name of the project seemed like a logical default suite as the name of the subprojects is used for tests within the subprojects. Also considered different "magic strings" such as "-" and just empty string, but those felt quite gimmicky as well. The used option parser also doesn't seem to support optional arguments for options (so that --suite could be given without a suite name).

Should the suite names actually be namespaced? Then you could have identically named suites within the main project and subprojects. Consider a project `foo` with a subproject `bar` and then you could invoke mesontest with suite arguments such as:

- foo/ # All tests in foo.
- foo/core
- foo/foo
- foo/helpers
- bar/ # All tests in bar.
- bar/core
- bar/bar

Maybe even support patterns:

- foo/util-*
- bar/featurex-*
- */core

The main project could be default "namespace" so that "foo/core" and "core" would mean the same thing.

**Comment by jpakkane** [comment]
:crossed_swords: conflicts.

There was a naming scheme that went something like `ninja test-subproj:suite` back when advanced tests were launched through ninja and not `mesontest`. That was not very nice because e.g. to run only main test set was done with `ninja test-`. Maybe we could add a new argument to mesontest, such as `--without-suite` which would run all tests except those mentioned?

There does not seem to be an immediate solution here. Would it be possible to commit this without the last patch and do that in a separate MR?

**Comment by trhd** [comment]
I has a plan. I'll open a new PR for the fifth commit.

**Comment by jpakkane** [comment]
Some test related changes got this to conflict. I'll hold off rest of the changes until we get this merged. Thanks.

**Pull request merged**